cmd/charm-ingest/whitelist.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/charmstore-client/internal/ingest"
+	"gopkg.in/juju/charmrepo.v4/csclient/params"
+)
+
+var validChannels = params.ValidChannels
+
+// parseWhitelist takes an io.Reader for a whitelist and returns []ingest.WhitelistEntity
+// or an error if unable to parse
+func parseWhitelist(r io.Reader) ([]ingest.WhitelistEntity, error) {
+	var entities []ingest.WhitelistEntity
+	fileScanner := bufio.NewScanner(r)
+	lineNum := 0
+
+	for fileScanner.Scan() {
+		lineNum++
+
+		entity, err := parseLine(fileScanner.Text(), lineNum)
+		if err != nil {
+			return nil, err
+		}
+
+		if entity.EntityId == "" {
+			// Empty line, ignore
+			continue
+		}
+		entities = append(entities, entity)
+	}
+
+	return entities, nil
+}
+
+// parseLine takes a string line of a whitelist file with its corresponding line number,
```

Comment:
// parseLine parses the given line into a WhitelistEntity. If the line is empty, it
    // returns the zero WhitelistEntity value.

?

In general we don't need to say the type of arguments or when we return an error, 'cos it should be evident from the signature.